java-checks/src/main/java/org/sonar/java/checks/AccessibilityChangeOnRecordsCheck.java

```diff
@@ -38,36 +37,45 @@
 import org.sonar.plugins.java.api.tree.Tree;
 
 @Rule(key = "S6216")
-public class AccessibilityChangeOnRecordsCheck extends AbstractAccessibilityChangeChecker implements JavaVersionAwareVisitor {
+public class AccessibilityChangeOnRecordsCheck extends AbstractAccessibilityChangeChecker {
   private static final String MESSAGE = "Remove this private field update which will never succeed";
   private static final String SECONDARY_MESSAGE = "Remove this accessibility bypass which will never succeed";
 
   private Map<Symbol, MethodInvocationTree> primaryTargets = new HashMap<>();
   private Map<Symbol, List<MethodInvocationTree>> secondaryTargets = new HashMap<>();
 
   @Override
-  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
-    return version.isJava16Compatible();
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.COMPILATION_UNIT, Tree.Kind.METHOD_INVOCATION);
   }
 
   @Override
-  public void leaveFile(JavaFileScannerContext context) {
-    for (Map.Entry<Symbol, MethodInvocationTree> entry : primaryTargets.entrySet()) {
-      Symbol symbol = entry.getKey();
-      MethodInvocationTree setInvocation = entry.getValue();
-      List<JavaFileScannerContext.Location> secondaries = secondaryTargets.getOrDefault(symbol, Collections.emptyList())
-        .stream()
-        .map(mit -> new JavaFileScannerContext.Location(SECONDARY_MESSAGE, mit))
-        .collect(Collectors.toList());
-      if (secondaries.isEmpty()) {
-        reportIssue(setInvocation, MESSAGE);
-      } else {
-        reportIssue(setInvocation, MESSAGE, secondaries, null);
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.COMPILATION_UNIT)) {
+      primaryTargets.clear();
+      secondaryTargets.clear();
+    }
+    super.visitNode(tree);
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(Tree.Kind.COMPILATION_UNIT)) {
+      for (Map.Entry<Symbol, MethodInvocationTree> entry : primaryTargets.entrySet()) {
+        Symbol symbol = entry.getKey();
+        MethodInvocationTree setInvocation = entry.getValue();
+        List<JavaFileScannerContext.Location> secondaries = secondaryTargets.getOrDefault(symbol, Collections.emptyList())
+          .stream()
+          .map(mit -> new JavaFileScannerContext.Location(SECONDARY_MESSAGE, mit))
+          .collect(Collectors.toList());
+        if (secondaries.isEmpty()) {
```

Comment:
This `if` doesn't seem necessary. You can use `reportIssue(setInvocation, MESSAGE, secondaries, null)` regardless of whether the list is empty or not.